chimney/src/main/scala/io/scalaland/chimney/internal/macros/TransformerMacros.scala

```diff
@@ -485,59 +499,68 @@ trait TransformerMacros extends TransformerConfigSupport with MappingMacros with
         Right(instanceTree)
       }
       .getOrElse {
-        val fromCS = From.typeSymbol.classSymbolOpt.get
-        val toCS = To.typeSymbol.classSymbolOpt.get
-
-        val fromInstances = fromCS.subclasses.map(_.typeInSealedParent(From))
-        val toInstances = toCS.subclasses.map(_.typeInSealedParent(To))
-
-        val targetNamedInstances = toInstances.groupBy(_.typeSymbol.name.toString)
-
-        val instanceClauses = fromInstances.map { instTpe =>
-          val instName = instTpe.typeSymbol.name.toString
-
-          resolveCoproductInstance(srcPrefixTree, instTpe, To, config)
-            .map { instanceTree =>
-              Right(cq"_: $instTpe => $instanceTree")
-            }
-            .getOrElse {
-              val instSymbol = instTpe.typeSymbol
-              targetNamedInstances.getOrElse(instName, Nil) match {
-                case List(matchingTargetTpe)
-                    if (instSymbol.isModuleClass || instSymbol.isCaseClass) && matchingTargetTpe.typeSymbol.isModuleClass =>
-                  val tree = mkTransformerBodyTree0(config) {
-                    q"${matchingTargetTpe.typeSymbol.asClass.module}"
-                  }
-                  Right(cq"_: ${instSymbol.asType} => $tree")
-                case List(matchingTargetTpe) if instSymbol.isCaseClass && matchingTargetTpe.typeSymbol.isCaseClass =>
-                  val fn = freshTermName(instName)
-                  expandDestinationCaseClass(Ident(fn), config.rec)(instTpe, matchingTargetTpe)
-                    .map { innerTransformerTree =>
-                      cq"$fn: $instTpe => $innerTransformerTree"
-                    }
-                case _ :: _ :: _ =>
-                  Left {
-                    Seq(
-                      AmbiguousCoproductInstance(
-                        instName,
-                        From.typeSymbol.fullName,
-                        To.typeSymbol.fullName
-                      )
-                    )
+        val fromInstances = getSealedHierarchyInstances(From)
+        val toInstances = getSealedHierarchyInstances(To)
+
+        val instanceClauses = fromInstances.flatMap {
+          case (canonicalName, instSymbols) =>
+            instSymbols.map { instSymbol =>
+              val instName = instSymbol.name.toString
+              val instTpe = instSymbol.typeInSealedParent(From)
+
+              resolveCoproductInstance(srcPrefixTree, instTpe, To, config)
+                .map { instanceTree =>
+                  Right(cq"_: $instTpe => $instanceTree")
+                }
+                .getOrElse {
+                  toInstances.getOrElse(canonicalName, Nil) match {
+                    case List(matchingTargetSymbol)
```

Comment:
@krzemin, I have attempted reworking this match clause so it's more readable. The idea is that we have 2 main branches: one where target can be expanded (case class), and where it's a singleton (object or java enum).

I decided to do this without much overengineering, and indeed it saved few lines of code. The other thought I had was to defining extractors and match on a pair `(instSymbol -> toInstances.getOrElse(canonicalName, Nil))` where cases would look like
```
case (JavaEnum(source), List(JavaEnum(target))) => ...
case (JavaEnum(source), List(CaseObject(target))) => ...
case (CaseObject(source), List(JavaEnum(target))) => ...
...
```

Can't decide which one would be better. Any suggestions?